internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,264 @@
+//
+// • jobRepoFn must be provided and is a function that returns the job repository
+//
+// • logger must be provided and is used to log errors that occur during scheduling and running of jobs
+//
+// WithRunJobsLimit and WithRunJobsInterval are the only valid options.
+func New(serverId string, jobRepoFn jobRepoFactory, logger hclog.Logger, opt ...Option) (*Scheduler, error) {
+	const op = "scheduler.New"
+	if serverId == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "missing server id")
+	}
+	if jobRepoFn == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing job repo function")
+	}
+	if logger == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing logger")
+	}
+
+	opts := getOpts(opt...)
+	return &Scheduler{
+		serverId:        serverId,
+		jobRepoFn:       jobRepoFn,
+		logger:          logger,
+		registeredJobs:  new(sync.Map),
+		runningJobs:     make(map[JobId]runningJob),
+		runJobsLimit:    opts.withRunJobsLimit,
+		runJobsInterval: opts.withRunJobInterval,
+	}, nil
+}
+
+// RegisterJob registers a job with the scheduler and persists the job into the repository.
+//
+// • job must be provided and is an implementer of the Job interface.
+//
+// • code must be provided and is used to uniquely distinguish between the same job implementation.
+//
+// The returned JobId is unique to the provide job and code.
+// All options are ignored.
+func (s *Scheduler) RegisterJob(ctx context.Context, j Job, code string, _ ...Option) (JobId, error) {
+	const op = "scheduler.(Scheduler).RegisterJob"
+	if err := validateJob(j); err != nil {
+		return "", errors.Wrap(err, op)
+	}
+	if code == "" {
+		return "", errors.New(errors.InvalidParameter, op, "missing code")
+	}
+
+	jobId, err := job.NewJobId(j.Name(), code)
+	if err != nil {
+		return "", errors.Wrap(err, op)
+	}
+
+	if _, ok := s.registeredJobs.Load(jobId); ok {
+		// Job already registered
+		return JobId(jobId), nil
+	}
+
+	repo, err := s.jobRepoFn()
+	if err != nil {
+		return "", errors.Wrap(err, op)
+	}
+
+	_, err = repo.CreateJob(ctx, j.Name(), code, j.Description())
+	if err != nil && !errors.IsUniqueError(err) {
+		return "", errors.Wrap(err, op)
+	}
+	s.registeredJobs.Store(JobId(jobId), j)
+
+	return JobId(jobId), nil
+}
+
+// UpdateJobNextRun sets the next scheduled run time for the provided jobId
+// to the current database time incremented by the nextRunIn parameter.  If
+// nextRunIn == 0 the job will be available to run immediately.
+//
+// • jobId must be provided and is the id of the job returned by RegisterJob
+//
+// All options are ignored.
+func (s *Scheduler) UpdateJobNextRun(ctx context.Context, jobId JobId, nextRunIn time.Duration, _ ...Option) error {
+	const op = "scheduler.(Scheduler).UpdateJobNextRun"
+	if jobId == "" {
+		return errors.New(errors.InvalidParameter, op, "missing job id")
+	}
+	_, ok := s.registeredJobs.Load(jobId)
```

Comment:
Is it possible that a jobId could be passed in here that was sourced in the DB elsewhere but has not yet been loaded into the cache?

Reply:
It shouldn't with the current setup but I guess it doesn't have to be a requirement here, lets chat tomorrow to walk through the process...